crates/relayer/src/chain/ckb4ibc.rs

```diff
@@ -398,27 +414,48 @@ impl Ckb4IbcChain {
             input_capacity,
             fee_rate,
         );
-        let (result, _) = self.rt.block_on(tx)?;
+        let (tx, extra_inputs) = self.rt.block_on(tx)?;
+
+        let total_inputs_capacity = extra_inputs
+            .into_iter()
+            .map(|v| Unpack::<u64>::unpack(&v.capacity()))
+            .sum::<u64>()
+            + input_capacity;
+        let total_outputs_capacity = tx.outputs_capacity().unwrap().as_u64();
+        assert!(
+            total_inputs_capacity > total_outputs_capacity,
+            "capacity overflow: {total_inputs_capacity} > {total_outputs_capacity}"
+        );
+
         let witness = WitnessArgs::new_builder()
-            .output_type(get_encoded_object(envelope).witness)
+            .output_type(get_encoded_object(&envelope).witness)
             .build()
             .as_bytes()
             .pack();
-        let result = result
+        let tx = tx
             .as_advanced_builder()
             // placeholder for the secp256k1 script, it will be used in the signing step
             .witness(WitnessArgs::new_builder().build().as_bytes().pack())
             .witness(witness)
             .build();
-        Ok(result)
+        Ok(tx)
+    }
+
+    fn counterparty_client_type(&self) -> ClientType {
+        self.counterparty_client_type
+            .borrow()
+            .unwrap_or(ClientType::Mock)
     }
 
     fn sync_counterparty_client_type(&self, client_type: ClientType) {
-        *self.counterparty_client_type.borrow_mut() = client_type;
-        if let Some(cmd) = &self.tx_internal_cmd {
-            cmd.send(client_type)
-                .expect("send counterparty_client_type");
-        }
+        self.counterparty_client_type.send_if_modified(|prev| {
+            if *prev != Some(client_type) {
+                *prev = Some(client_type);
```

Comment:
``` rust
assert!(prev.is_none());
```
Ensure the client_type isn't set.

Reply:
`prev` can be Some or None, why should ensure it always be the None?